app/scripts/metamask-controller.js

```diff
@@ -1253,6 +1263,9 @@ module.exports = class MetamaskController extends EventEmitter {
     engine.push(createLoggerMiddleware({ origin }))
     engine.push(filterMiddleware)
     engine.push(this.preferencesController.requestWatchAsset.bind(this.preferencesController))
+    engine.push(this.createTypedDataMiddleware('eth_signTypedData', 'V1').bind(this))
+    engine.push(this.createTypedDataMiddleware('eth_signTypedData_v1', 'V1').bind(this))
+    engine.push(this.createTypedDataMiddleware('eth_signTypedData_v2', 'V2').bind(this))
```

Comment:
Instead of using `processTypedMessage` when initializing the web3-provider-engine, we use custom middleware to intercept `eth_signTypedData` methods. This allows us to have fine-grained control over custom and non-custom RPC methods without touching the provider engine itself.

Note that for now, `eth_signTypedData` is still pointing to `V1` of the EIP-712 proposal.

Reply:
I told Pete Kim of Cipher that his implementation could be v2, and so the final one (this one) would be `v3`. Mind if we change to that? I think that will be a nice gesture to them, and leaves us open to backwards-support any apps using their version.